client/client.go

```diff
@@ -1,19 +1,307 @@
+}
+
+// Defines the Ownership Voucher format. See https://www.rfc-editor.org/rfc/rfc8366.html.
+type OwnershipVoucherInner struct {
+	CreatedOn                  string `json:"created-on"`
+	ExpiresOn                  string `json:"expires-on"`
+	SerialNumber               string `json:"serial-number"`
+	Assertion                  string `json:"assertion"`
+	PinnedDomainCert           string `json:"pinned-domain-cert"`
+	DomainCertRevocationChecks bool   `json:"domain-cert-revocation-checks"`
+}
+
+// pemEncodeCert adds the correct PEM headers and footers to a raw certificate block.
+func pemEncodeCert(contents string) string {
+	return "-----BEGIN CERTIFICATE-----\n" + contents + "\n-----END CERTIFICATE-----"
 }
 
-func (c *Client) GetBootstrapData(ctx context.Context, req *bootz.GetBootstrapDataRequest, opts ...grpc.CallOption) (*bootz.GetBootstrapDataResponse, error) {
+// validateArtifacts checks the signed artifacts in a GetBootstrapDataResponse. Specifically, it:
+// - Checks that the OV in the response is signed by the manufacturer
+// - Checks that the serial number in the OV matches the one in the original request.
+// - Verififies that the Ownership Certificate is in the chain of signers of the Pinned Domain Cert.
+func validateArtifacts(serialNumber string, resp *bootz.GetBootstrapDataResponse, rootCA []byte) error {
+	ov64 := resp.GetOwnershipVoucher()
+	if len(ov64) == 0 {
+		return fmt.Errorf("received empty ownership voucher from server")
+	}
+
+	oc := resp.GetOwnershipCertificate()
+	if len(oc) == 0 {
+		return fmt.Errorf("received empty ownership certificate from server")
+	}
+	// Decode the ownership voucher
+	ov, err := base64.StdEncoding.DecodeString(string(ov64))
+	if err != nil {
+		return err
+	}
+
+	// Parse the PKCS7 message
+	p7, err := pkcs7.Parse(ov)
+	if err != nil {
+		return err
+	}
+	// Unmarshal the ownership voucher into a struct.
+	parsedOV := OwnershipVoucher{}
+	err = json.Unmarshal(p7.Content, &parsedOV)
+	if err != nil {
+		return err
+	}
+
+	// Create a CA pool for the device to validate that the vendor has signed this OV.
+	vendorCAPool := x509.NewCertPool()
+	if !vendorCAPool.AppendCertsFromPEM(rootCA) {
+		return fmt.Errorf("unable to add vendor root CA to pool")
+	}
+
+	// Verify the ownership voucher with this CA.
+	err = p7.VerifyWithChain(vendorCAPool)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Validated ownership voucher signed by vendor")
+
+	// Verify the serial number for this OV
+	if parsedOV.OV.SerialNumber != serialNumber {
+		return fmt.Errorf("serial number from OV does not match request")
+	}
+
+	pdCPEM := pemEncodeCert(parsedOV.OV.PinnedDomainCert)
+
+	// Create a new pool with this PDC.
+	pdcPool := x509.NewCertPool()
+	if !pdcPool.AppendCertsFromPEM([]byte(pdCPEM)) {
+		return err
+	}
+
+	// Parse the Ownership Certificate.
+	ocCert, err := certFromPemBlock(oc)
+	if err != nil {
+		log.Exitf("failed to parse certificate: %v", err)
```

Comment:
shouldn't exit from non main functions just return the error and let caller decide what to do

Reply:
Thanks, missed this one.